src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -130,6 +156,55 @@ public ImmutableArray<PackageSource> GetPackageSources()
             return packageSources;
         }
 
+        private async Task<ImmutableArray<PackageSource>> GetPackageSourcesImplAsync()
+        {
+            // Automatically retry the operation if it gets cancelled due to a change in the workspace.
+            while (true)
+            {
+                var cancellationToken = _tokenSource.Token;
+                try
+                {
+                    var result = await GetPackageSourcesImplAsync(cancellationToken).ConfigureAwait(false);
+                    if (result.IsEmpty && cancellationToken.IsCancellationRequested)
```

Comment:
> Why are we handling both this and also the OperationCancelledException in the catch block?

Cancelling a token may result in an exception, but this is never a guarantee (and rarely a guarantee even in cases where the token is cancelled before the start of the call). Cancellation tokens are frequently checked after calls.

> if the only thing actually using the cancellationToken is the SwitchToMainThreadAsync() then I don't see how this could happen in a way we'd care about?

I don't understand this part of the comment. Can you rephrase?

Reply:
> Cancellation tokens are frequently checked after calls.

I guess I'd prefer it was just cancellationToken.ThrowIfCancellationRequested() which can jump to the catch block; that way you're not duplicating the retry logic.

Reply:
> if the only thing actually using the cancellationToken is the SwitchToMainThreadAsync() then I don't see how this could happen in a way we'd care about?

As you observed on my other question the only way we can hit this is if it failed, and then we had the badly timed workspace change. But why does that combination require special handling like this? I would have assumed either:

1. If it fails, it just failed, and whether we failed to observe the cancellation is moot -- it's broken either way. Either we accept the brokenness or want to retry, period.
2. We had cancelled, but if we got back a successful result, why not just use it? We didn't want it, but if you've already done the work why not keep it?

Reply:
Removed the special case handling here.